### ISSUE TITLE
JPEG: fails to decode some images

### ISSUE THREAD

**Issue by Shnatsel** [etemesi254/zune-image#55]
I've found that my testing script wasn't capturing decoding failures properly. It checked for decoding mismatches correctly, but the failure of the `zune` tool to decode an image was not recorded correctly.

I've fixed it and re-ran the decoding test, and found that ~300 images out of ~40,000 tested were not correctly decoded by the `zune` tool. The archive containing them can be found [here](https://mega.nz/file/4tVnCTxb#_sFsOdFEBBZ-7pQO8YZTbhTaJt81eYpZPtC4GpkoIQM).

Tested on commit 74c36b8aba79ebde04fd43ddd92afb73198c9d5b

**Comment by etemesi254** [comment]
Some interesting images there.

Do the licenses allow me to add some into test corpus?

**Comment by Shnatsel** [comment]
I can't really speak for 300 images from disparate sources all at once. 

Legally the default is that everything is copyrighted, so e.g. Debian won't include these images in their packages. In practice it's overwhelmingly likely that nobody will mind or care.

FWIW I suggest storing them outside the git repository to avoid bloating it, since it has to include everything that was ever committed to it, and large images can inflate the download size for just cloning the repo quite drastically. I suggest using some other solution; e.g. Github provides [git large file storage](https://docs.github.com/en/repositories/working-with-files/managing-large-files) to help deal with this.

**Comment by etemesi254** [comment]
git lfs limits one to 1 gb of bandwidth and 1gb of storage per month.

I have no problem with keeping storage in check but I fear bandwidth may run out, with how the CI's are set since they fuzz per day and run tests per PR/push.

**Comment by Shnatsel** [comment]
Fuzzing should be done on small files, under 1Kb. If you're feeding large JPEGs to the fuzzer, it won't find much in the way of interesting codepaths.

You can either use the [AFL](https://lcamtuf.coredump.cx/afl/demo/) corpora to kickstart the fuzzer, or better yet I can share mine that have been accumulated over time and over fuzzing several JPEG implementations.

**Comment by etemesi254** [comment]
I'd love to have your corpus.

Ideally for all formats that you currently have.

**Comment by etemesi254** [comment]
An update on this was that most were fixed and we added support for some arcane jpeg images with interesting colorspaces (CYMK and YCCK), there is still some few that don't work. So this can't be closed yet, but it's close.

Also pinging this to ask about the fuzz corpus, I was thinking it's better you set it up since you do a lot more fuzzing than I do and more crates can benefit from this.

But there is no hurry.

Can't wait to close this and declare zune-jpeg Shnatsel approved

**Comment by etemesi254** [comment]
Could you rerun this, most should be fixed and tests to ensure they don't happen again added

**Comment by Shnatsel** [comment]
I've re-run the test, I'll process the results and post the failures (if any) in the next few days.

**Comment by Shnatsel** [comment]
There are still decode failures on 47 images: [zune-jpeg-failures-2.tar.gz](https://github.com/etemesi254/zune-image/files/10451091/zune-jpeg-failures-2.tar.gz)

I haven't looked at divergences, I'll re-run that test and look at that once the decode failures are gone.

**Comment by etemesi254** [comment]
I have confirmed that all images in the above decode correctly except the two tiff files attached as of 9ac4184e0c2a528f066035d4aab4e7a44e4e0c3e.

Additionally some were added to the test suite to ensure no regressions occur

**Comment by Shnatsel** [comment]
There's just one failing image remaining as of 48ac1d1d85068683cc5f7ae0d71c67dbc5813c1f (the commit doesn't build, I had to fix up the QOI error reporting).

![In%20Pursuit,%20Cheetahs](https://user-images.githubusercontent.com/291257/213884946-cb492f1f-27e8-4b72-bc62-77fb4470ad1c.jpg)

The error message is: 

```
ERROR [zune_bin]  Could not complete workflow, reason jpg: "Corrupt Adobe App14 segment"
```

**Comment by Shnatsel** [comment]
No new divergences in the decoding results for the images that were decoded successfully.

**Comment by etemesi254** [comment]
Fixed as of 39a4d3f14147d2532b6cfe9c982af99702e345f5


Sorry for the compilation errors, found me as I was doing weekend updates(adding qoi encoding to the command line)

**Comment by Shnatsel** [comment]
Then I'm going to go ahead and close this. Yay!
